### PR TITLE
api: PutObjectSignV2 should support uploading 0byte objects.

### LINKED ISSUE #761: Allow 0byte files to be uploaded with PutObject() with SignatureV2
Current master fails to upload 0 byte files while using signature v2.  Thanks to @sb10 for reproducing and reporting the problem. 
```
touch /tmp/a
go run putobject.go
```

```go

package main

import (
	"log"
	"os"

	minio "github.com/minio/minio-go"
)

func main() {
	s3Client, err := minio.NewV2("s3.amazonaws.com", os.Getenv("ACCESS_KEY"), os.Getenv("SECRET_KEY"), false)
	if err != nil {
		log.Fatalln(err)
	}
	s3Client.TraceOn(os.Stderr)
	n, err := s3Client.FPutObject("my-bucketname", "myobject", "/tmp/a", "")
	if err != nil {
		log.Fatalln(err)
	}
	log.Println("Uploaded", "my-objectname", " of size: ", n, "Successfully.")
}
```

Error returned by AWS S3
```
---------START-HTTP---------
POST /myobject?uploadId=OQUU.mQ835gwBd2b6XiNcQgf2ceUPyuNITzNIV9dJC2MLJq2.AM0q16Wb2C6q5LXIye4kWK57pz05YkHch5xW1B5w0cEPOiTs_MyafUT5uAd88R_TM77vP9HoU4RJ05F HTTP/1.1
Host: my-bucketname.s3-us-west-1.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/3.0.0
Content-Length: 99
Authorization: AWS **REDACTED**:**REDACTED**
Date: Fri, 14 Jul 2017 16:12:38 GMT
Accept-Encoding: gzip

HTTP/1.1 400 Bad Request
Connection: close
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Fri, 14 Jul 2017 16:12:39 GMT
Server: AmazonS3
X-Amz-Id-2: /ZfYlcbc+GdiOQ3kbggw2vwY5RdN2qaI8kCDMWacSo5n4nHtsPM6+HQgP4FOw6Nu+z65Hch5w24=
X-Amz-Request-Id: 0B9D1005AB31FA62

118
<Error><Code>MalformedXML</Code><Message>The XML you provided was not well-formed or did not validate against our published schema</Message><RequestId>0B9D1005AB31FA62</RequestId><HostId>/ZfYlcbc+GdiOQ3kbggw2vwY5RdN2qaI8kCDMWacSo5n4nHtsPM6+HQgP4FOw6Nu+z65Hch5w24=</HostId></Error>
0
---------END-HTTP---------
---------START-HTTP---------
DELETE /myobject?uploadId=OQUU.mQ835gwBd2b6XiNcQgf2ceUPyuNITzNIV9dJC2MLJq2.AM0q16Wb2C6q5LXIye4kWK57pz05YkHch5xW1B5w0cEPOiTs_MyafUT5uAd88R_TM77vP9HoU4RJ05F HTTP/1.1
Host: my-bucketname.s3-us-west-1.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/3.0.0
Authorization: AWS **REDACTED**:**REDACTED**
Date: Fri, 14 Jul 2017 16:12:38 GMT
Accept-Encoding: gzip

Date: Fri, 14 Jul 2017 16:12:40 GMT
Server: AmazonS3
X-Amz-Id-2: jE0dSKAdUoF8pciXeM9zXqz1XqYNnfGcqubx1YXzfbpyGwBwq1wuVvKvKrg0R4mnGw0EHC4dy5A=
X-Amz-Request-Id: C40CDB9CBC998303
---------END-HTTP---------
2017/07/14 09:12:38 The XML you provided was not well-formed or did not validate against our published schema
exit status 1
```

### PR DESCRIPTION
**Pull request by harshavardhana**
Fixes #761

### CODE REVIEW

**Comment by harshavardhana** [comment]
Tests have been added @vadmeste @krisis  - please review.

File: api-put-object-multipart.go [hunk 1]
```diff
@@ -141,7 +141,7 @@ func (c Client) putObjectMultipartNoStream(bucketName, objectName string, reader
 	}
 
 	// Verify if we uploaded all the data.
-	if size > 0 {
+	if size >= 0 {
```
Comment (krisis) [inline api-put-object-multipart.go]: `putObjectNoChecksum` is called when `size` is zero. So, this check is not needed.

File: api-put-object-common.go [hunk 2]
```diff
@@ -58,6 +58,10 @@ func optimalPartInfo(objectSize int64) (totalPartsCount int, partSize int64, las
 	if objectSize == -1 {
 		objectSize = maxMultipartPutObjectSize
 	}
+	// object size is '0' set default parts.
+	if objectSize == 0 {
+		return 1, 0, 0, nil
+	}
```
Comment (krisis) [inline api-put-object-common.go]: `putObjectNoChecksum` is called when `size` is zero. So, this check is not needed.

File: functional_tests.go [hunk 3]
```diff
@@ -3596,6 +3596,64 @@ func testUserMetadataCopyingV2() {
 	testUserMetadataCopyingWrapper(c)
 }
 
+// Test put object with 0 byte object.
```
Comment (krisis) [inline functional_tests.go]: We should have a test for V4 too.

Comment (harshavardhana) [inline functional_tests.go]: V4 we already have PutObjectStreaming uploads a 0byte file.

File: api-put-object-common.go (hunk 2 above)
Comment (harshavardhana) [inline api-put-object-common.go]: This code should work accordingly regardless of how its stitched together

File: api-put-object-multipart.go (hunk 1 above)
Comment (harshavardhana) [inline api-put-object-multipart.go]: The reason this is done is multipart should also work for size == 0 independent of its upper layer usage.

**Comment by harshavardhana** [comment]
@vadmeste can you take a look PR is updated with unit test as requested.

**Pull request merged**